Rebuild Docs V0.8.0 (#8392)
Oneflow-Inc/oneflow: docs/source/distributed.rst

oneflow.distributed
=========================================================

.. note ::
    Please refer to `OneFlow Distributed Overview <https://docs.oneflow.org/master/parallelism/01_introduction.html>`__
    for a brief introduction to all features related to distributed training.

OneFlow provides two ways to accomplish `Distributed Training`:

> - One of them is to use the original concept of OneFlow to run data parallel training by configurating global tensor，introduces the concept of **Global View** to simplify distributed training. Simply , in OneFlow's global view, a cluster is abstracted as a "supercomputing device".

Reviewer: 不只是`data parallel training`，另外总觉得这段话表达的不清楚，有点绕口。我写一段你参考一下哈，
```
- The first way is that users are recommended to use OneFlow's global Tensor for distributed training. Global Tensor regards the computing cluster as a supercomputing device, allowing users to write distributed training code just like in a single-machine environment.
```

> - Besides, to facilitate the users who are transferring from PyTorch to OneFlow, OneFlow provide the interface consistent with PyTorch ``torch.nn.parallel.DistributedDataParallel``, ``oneflow.nn.parallel.DistributedDataParallel``.

Reviewer: 这里也稍微改一下，不要在文档中透露出我们要让用户从PyTorch迁移到OneFlow的意图。
```
- OneFlow also provides a DDP module aligned with PyTorch. DDP has been well-known and widely used in data parallelism by the majority of PyTorch users. Also see [PyTorch DDP introduction]().
```

Author: 已改

Baisc
-------------------------------
When you start distributed training in OneFlow,the following functions can be used.

.. currentmodule:: oneflow.env

.. autosummary::
    :toctree: generated
    :nosignatures:

    get_world_size
    get_rank
    get_local_rank
    get_node_size
    init_rdma
    rdma_is_initialized

> DistributedDataParallel

Reviewer: 这里顺序是不是有点反了，前面说两种方式的时候是先说的Global Tensor，后说的DDP

        y_pred = m(x)
        l = loss(y_pred, y)
        if (i + 1) % 50 == 0:
            print(f"{i+1}/{m.iter_count} loss:{l}")

        optimizer.zero_grad()
        l.backward()
        optimizer.step()

    print(f"\nw:{m.w}")

There are only two differences between the data parallelism training code and the stand-alone single-card script:

- Use `DistributedDataParallel` to wrap the module object (`m = ddp(m)`)
- Use `get_rank` to get the current device number and distribute the data to the device.

Then use `launcher` to run the script, leave everything else to OneFlow, which makes distributed training as simple as stand-alone single-card training:

::

    python3 -m oneflow.distributed.launch --nproc_per_node 2 ./ddp_train.py

Communication collectives
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
.. currentmodule:: oneflow.comm
    
.. autosummary::
    :toctree: generated
    :nosignatures:

        all_reduce
        all_gather
        all_to_all
        broadcast
        barrier
        gather
        reduce
        reduce_scatter
        recv
        scatter
        send

Global Tensor
--------------------------------------------------------------

> Create it

Author: 已改

^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Global Tensor is a Tensor designed to facilitate distributed execution on multiple machines and devices, and is an interface for implementing Global View programming.

> In each of the two consoles, import oneflow and create x

Author: 已修改

::

    >>>import oneflow as flow
    >>> # Place a global tensor on cuda device of rank(process) 0 and 1
    >>> placement = flow.placement(type="cuda", ranks=[0, 1])
    >>> # Each rank's local data is a part data as a result of spliting global data on dim 0
    >>> sbp = flow.sbp.split(dim=0)
    >>> # Create a global tensor by randn
    >>> x = flow.randn(4, 5, placement=placement, sbp=sbp)
    >>> x.shape
    oneflow.Size([4, 5])

> Get Local Tensor from Global Tensor

Author: 已改

^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

> Call `to_local()` to check the local tensor on a device.

Author: 已改

::

    x.to_local()
    tensor([[ 2.9186e-01, -3.9442e-01,  4.7072e-04, -3.2216e-01,  1.7788e-01],
            [-4.5284e-01,  1.2361e-01, -3.5962e-01,  2.6651e-01,  1.2951e+00]],

>         device='cuda:0', dtype=oneflow.float32)

Reviewer 2: 可以写成
```
y = x.to_local()
y.is_local
True
y
tensor([[ 2.9186e-01, -3.9442e-01,  4.7072e-04, -3.2216e-01,  1.7788e-01],
            [-4.5284e-01,  1.2361e-01, -3.5962e-01,  2.6651e-01,  1.2951e+00]],
        device='cuda:0', dtype=oneflow.float32)

```

Author: 已改

Convert Local Tensor to Global Tensor
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

Developers can create local tensor first, then convert it to global tensor with `Tensor.to_global`.

> Two local tensors with the shape of ``(2,5)`` are created separately on two devices. While after the to_global method, the global tensor with a shape of ``(4,5)`` is obtained.

Reviewer: 这几个to_local\to_global接口的描述性文字参考一下0.7的release note（https://github.com/Oneflow-Inc/oneflow/releases ）再重写一下吧

Author: 已修改

oneflow.distributed
=========================================================

.. note ::
    Please refer to `OneFlow Distributed Overview <https://docs.oneflow.org/master/parallelism/01_introduction.html>`__
    for a brief introduction to all features related to distributed training.

OneFlow provides two ways to accomplish `Distributed Training`:

- The first way is that users are recommended to use OneFlow's global Tensor for distributed training. Global Tensor regards the computing cluster as a supercomputing device, allowing users to write distributed training code just like in a single-machine environment.

- OneFlow also provides a DDP（DistributedDataParallel） module aligned with PyTorch. DDP has been well-known and widely used in data parallelism by the majority of PyTorch users. Also see `PyTorch DDP introduction` <https://pytorch.org/docs/stable/generated/torch.nn.parallel.DistributedDataParallel.html#torch.nn.parallel.DistributedDataParallel>_.

Baisc
-------------------------------

> When you start distributed training in OneFlow,the following functions can be used.

Reviewer: ```suggestion
When you start distributed training in OneFlow, the following functions can be used.
```

.. currentmodule:: oneflow.env

.. autosummary::
    :toctree: generated
    :nosignatures:

    get_world_size
    get_rank
    get_local_rank
    get_node_size
    init_rdma
    rdma_is_initialized

Global Tensor
--------------------------------------------------------------

Build/Construct Global Tensor
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
A global Tensor can be created with a ``placement`` and a ``sbp``. The ``placement`` describes the physical devices of the global tensor will be allocated, and the ``sbp`` describes its distribution along these devices.

::

    >>>import oneflow as flow
    >>> # Place a global tensor on cuda device of rank(process) 0 and 1
    >>> placement = flow.placement(type="cuda", ranks=[0, 1])
    >>> # Each rank's local data is a part data as a result of spliting global data on dim 0
    >>> sbp = flow.sbp.split(dim=0)
    >>> # Create a global tensor by randn
    >>> x = flow.randn(4, 5, placement=placement, sbp=sbp)
    >>> x.shape
    oneflow.Size([4, 5])

> Convert Local Tensor to Global Tensor

Author: 有对应的参考资料么，用户文档和 0.7 的 release 版本都没有找到相关的介绍

> Convert Local Tensor to Global Tensor

Reviewer: 里面Supporting redistribution of Global Tensor in clusters就是

^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

With ``Tensor.to_global`` interface, you can create a `Global Tensor` based on a `Local Tensor`, and regard this tensor as the local tensor of the `Global Tensor`` on the present device.

Two local tensors with the shape of ``(2,5)`` are created separately on two devices. While after the to_global method, the global tensor with a shape of ``(4,5)`` is obtained.

Code running on Node 0

::

    import oneflow as flow

    x = flow.randn(2,5)
    placement = flow.placement("cuda", [0,1])
    sbp = flow.sbp.split(0)
    x_global = x.to_global(placement=placement, sbp=sbp)
    x_global.shape

Code running on Node 1

::

    import oneflow as flow

    x = flow.randn(2,5)
    placement = flow.placement("cuda", [0,1])
    sbp = flow.sbp.split(0)
    x_global = x.to_global(placement=placement, sbp=sbp)
    x_global.shape

Get Local Tensor from Global Tensor
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

> With ``Tensor.to_local`` interface, you can return the `local tenso`r of the `Global Tensor` on the present device.

Reviewer: ```suggestion
With ``Tensor.to_local`` interface, you can return the local tensor of the `Global Tensor` on the present device.
```

oneflow.distributed
=========================================================

.. note ::
    Please refer to `oneflow Distributed Overview <https://docs.oneflow.org/master/parallelism/01_introduction.html>`__
    for a brief introduction to all features related to distributed training.

oneflow provides two ways to accomplish `Distributed Training`:

> - One of them is to use the original concept of Oneflow to run data parallel training by configurating global tensor，introduces the concept of **Global View** to simplify distributed training. Simply , in OneFlow's global view, a cluster is abstracted as a "supercomputing device".

Author: 已修改

> - Besides, to facilitate the users who are transferring from PyTorch to OneFlow, OneFlow offers the interface consistent with PyTorch ``torch.nn.parallel.DistributedDataParallel``, ``oneflow.nn.parallel.DistributedDataParallel``.

Author: 已修改

Baisc
-------------------------------

> When you start distributed training in oneflow,the following functions can be used.

Author: 已修改

OneFlow provides two ways to accomplish `Distributed Training`:

- The first way is that users are recommended to use OneFlow's global Tensor for distributed training. Global Tensor regards the computing cluster as a supercomputing device, allowing users to write distributed training code just like in a single-machine environment.

- OneFlow also provides a DDP（DistributedDataParallel） module aligned with PyTorch. DDP has been well-known and widely used in data parallelism by the majority of PyTorch users. Also see `PyTorch DDP introduction <https://pytorch.org/docs/stable/generated/torch.nn.parallel.DistributedDataParallel.html#torch.nn.parallel.DistributedDataParallel>`_.

Baisc
-------------------------------
When you start distributed training in OneFlow, the following functions can be used.

.. currentmodule:: oneflow.env

.. autosummary::
    :toctree: generated
    :nosignatures:

    get_world_size
    get_rank
    get_local_rank
    get_node_size
    init_rdma
    rdma_is_initialized

Global Tensor
--------------------------------------------------------------

Construct Global Tensor
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

> A global Tensor can be created with a ``placement`` and a ``sbp``. The ``placement`` describes the physical devices of the global tensor will be allocated, and the ``sbp`` describes its distribution along these devices.

Reviewer: ```suggestion
A global Tensor can be created with a ``placement`` and a ``sbp``. The ``placement`` describes the physical devices of the global tensor will be allocated, and the ``sbp`` describes its distribution among these devices.
```

With ``Tensor.to_global`` interface, you can create a `Global Tensor` based on a `Local Tensor`, and regard this tensor as the local tensor of the `Global Tensor` on the present device.

Two local tensors with the shape of ``(2,5)`` are created separately on two devices. While after the to_global method, the global tensor with a shape of ``(4,5)`` is obtained.

Code running on Node 0

::

    import oneflow as flow

    x = flow.randn(2,5)
    placement = flow.placement("cuda", [0,1])
    sbp = flow.sbp.split(0)
    x_global = x.to_global(placement=placement, sbp=sbp)
    x_global.shape

Code running on Node 1

::

    import oneflow as flow

    x = flow.randn(2,5)
    placement = flow.placement("cuda", [0,1])
    sbp = flow.sbp.split(0)
    x_global = x.to_global(placement=placement, sbp=sbp)
    x_global.shape

Redistribute Global Tensor
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
With ``Tensor.to_global`` interface, data of `Global Tensor` can be redistributed the in clusters,either by choosing to distribute to another set of nodes or by changing its distribution on that set of nodes (i.e. changing the SBP)

> Redistribution of data usually occurs with cross-process data communication, and the ``Tensor.to_global`` interface finely shields the complex underlying communication logic.

Author: 已删除

::

    >>> import oneflow as flow
    >>> x = flow.tensor([1.0, 2.0], placement=flow.placement("cuda", ranks=[0, 1]), sbp=flow.sbp.split(0))
    >>> y = x.to_global(placement=flow.placement("cuda", ranks=[2, 3]), sbp=flow.sbp.broadcast)

> Each operator of OneFlow defines a set of combinations of input and output SBPs that it can support, and `Global Tensor`` supports automatic redistribution to meet the requirements of executing a particular computational interface on its SBP. For example, the following code:

Author: 已修改

::

    >>> import oneflow as flow
    >>> x = flow.randn(4, 4, 
            placement=flow.placement("cuda", ranks=[0, 1]), 
            sbp=flow.sbp.split(0))
    >>> y = flow.randn(4, 4, 
            placement=flow.placement("cuda", ranks=[0, 1]), 
            sbp=flow.sbp.split(1))
    >>> z = x + y

> When ``x + y`` is executed, since x is cut by dimension ``0`` and y is cut by dimension ``1``, their components at each node cannot be added directly, so it automatically converts the SBP of x to ``flow.sbp.split(1)`` or y to ``flow.sbp.split(0)``, and the SBP of z is calculated as ``flow.sbp. split(1)`` or ``flow.sbp.split(0)``.

Author: 已修改

.. note ::

>     - Global Tensor does not currently support mixing with the DDP interface.

Author: 已修改

>     - Global Tensor requires all devices to execute simultaneously, and the code that has branches would lead to process deadlock because of divergent execution paths. We will continue to improve the user experience here.

Reviewer: ```suggestion
    - Global Tensor requires all devices to execute at the same pace, otherwise, it may cause multi-process deadlock.
```

Author: 已修改

    for a brief introduction to all features related to distributed training.

OneFlow provides two ways to accomplish `Distributed Training`:

- The first way is that users are recommended to use OneFlow's global Tensor for distributed training. Global Tensor regards the computing cluster as a supercomputing device, allowing users to write distributed training code just like in a single-machine environment.

- OneFlow also provides a DDP（DistributedDataParallel） module aligned with PyTorch. DDP has been well-known and widely used in data parallelism by the majority of PyTorch users. Also see `PyTorch DDP introduction <https://pytorch.org/docs/stable/generated/torch.nn.parallel.DistributedDataParallel.html#torch.nn.parallel.DistributedDataParallel>`_.

Baisc
-------------------------------
When you start distributed training in OneFlow, the following functions can be used.

.. currentmodule:: oneflow.env

.. autosummary::
    :toctree: generated
    :nosignatures:

    get_world_size
    get_rank
    get_local_rank
    get_node_size
    init_rdma
    rdma_is_initialized

Global Tensor
--------------------------------------------------------------

> Build/Construct Global Tensor

Reviewer 2: 这里的Build和Construct选一个就行，我的评论不是要你两个都写 😂

Author: 已改

    >>> sbp = flow.sbp.split(dim=0)
    >>> # Create a global tensor by randn
    >>> x = flow.randn(4, 5, placement=placement, sbp=sbp)
    >>> x.shape
    oneflow.Size([4, 5])

Convert Local Tensor to Global Tensor
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

With ``Tensor.to_global`` interface, you can create a `Global Tensor` based on a `Local Tensor`, and regard this tensor as the local tensor of the `Global Tensor` on the present device.

Two local tensors with the shape of ``(2,5)`` are created separately on two devices. While after the to_global method, the global tensor with a shape of ``(4,5)`` is obtained.

Code running on Node 0

::

    import oneflow as flow

    x = flow.randn(2,5)
    placement = flow.placement("cuda", [0,1])
    sbp = flow.sbp.split(0)
    x_global = x.to_global(placement=placement, sbp=sbp)
    x_global.shape

Code running on Node 1

::

    import oneflow as flow

    x = flow.randn(2,5)
    placement = flow.placement("cuda", [0,1])
    sbp = flow.sbp.split(0)
    x_global = x.to_global(placement=placement, sbp=sbp)
    x_global.shape

Redistribute Global Tensor
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

> With ``Tensor.to_global`` interface, data of `Global Tensor` can be redistributed the in clusters,either by choosing to distribute to another set of nodes or by changing its distribution on that set of nodes (i.e. changing the SBP)

Reviewer: 这段和下面那段重复了吧

Author: 已改

oneflow.distributed
=========================================================

.. note ::
    Please refer to `oneflow Distributed Overview <https://docs.oneflow.org/master/parallelism/01_introduction.html>`__
    for a brief introduction to all features related to distributed training.

OneFlow introduces the concept of **Global View** to simplify distributed training. Simply , in OneFlow's global view, a cluster is abstracted as a "supercomputing device".

> OneFlow's **Global View**  relies on several important concepts: ``Placement``, ``SBP`` and ``SBP Signature``.

Reviewer: 我认为SBP Signature可以删了，从头到尾就只有``Placement``和``SBP``，SBP Signature实际上和SBP就是一个东西，placement也没有placement signature这一说法

> OneFlow's **Global View**  relies on several important concepts: ``Placement``, ``SBP`` and ``SBP Signature``.

Author: 好的，我将做一个概述，并分出一个大标题的 ddp ，对标 pytorch 的 ddp 进行简要介绍，另一个大标题介绍 global

Basics
--------------------------------------------------------------

Placement
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

> The Tensor in the OneFlow global view has a placement attribute, which allows you to specify the physical device on which the Tensor is stored.

Reviewer: 我记得Placement和SBP在TensorAttributes里有对应的API文档，这里还需要重新写一遍么，而且两个地方说法还有区别。https://oneflow.readthedocs.io/en/master/tensor_attributes.html

Author: 1、好的，我之前也有疑惑 SBP Signature 是否要加，因为没有对应的接口表示，将会删除。

2、由于我是想将 Placement 和 SBP 作为 oneflow 特有的模块进行介绍，所以想将这些接口放在 oneflow.distributed 模块，并将会同时删除在 tensorAttribute里对应的接口文档，并且 TensorAttribute 文档里也丰富了内容  https://github.com/Oneflow-Inc/oneflow/blob/ca8da95240ce6e19af72acf1276dfbcd156ec258/docs/source/tensor_attributes.rst

.. autoclass:: oneflow.placement

SBP
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

SBP is a concept invented by OneFlow that describes the mapping of data in the global view of a "supercomputing device" to data on real physical devices in a cluster, and is a combination of the initials split, broadcast, and partial.

.. autoclass::  oneflow.sbp.sbp

SBP Signature
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

Create global tensor
--------------------------------------------------------------
Global Tensor is a Tensor designed to facilitate distributed execution on multiple machines and devices, and is an interface for implementing Global View programming.

In each of the two consoles, import oneflow and create x

::

    >>>import oneflow as flow
    >>> # Place a global tensor on cuda device of rank(process) 0 and 1
    >>> placement = flow.placement(type="cuda", ranks=[0, 1])
    >>> # Each rank's local data is a part data as a result of spliting global data on dim 0
    >>> sbp = flow.sbp.split(dim=0)
    >>> # Create a global tensor by randn
    >>> x = flow.randn(4, 5, placement=placement, sbp=sbp)
    >>> x.shape
    oneflow.Size([4, 5])

> 

Reviewer: 前面那些要不都删了吧，这个只作为DDP相关接口的文档，binbin意下如何？ @clackhan

Reviewer 2: > 前面那些要不都删了吧，这个只作为DDP相关接口的文档，binbin意下如何？ @clackhan

可以，专门一个模块介绍global，这里只介绍DDP

Post-Initialization
--------------------------------------------------------------
.. currentmodule:: oneflow.env

.. autosummary::
    :toctree: generated
    :nosignatures:

    get_world_size
    get_rank
    get_local_rank
    get_node_size
    init_rdma
    rdma_is_initialized

Communication collectives
--------------------------------------------------------------
.. currentmodule:: oneflow.comm
    

> .. autosummary::

Reviewer 2: 这里可以写详细一点，包括解释和用法
